samples: net/vlan: add IPv6 prefix config to each vlan-iface

shrek-wang opened the pull request:
Without the IPv6-prefix set, the net_if_ipv6_addr_onlink() check will always fail. The packets with DIP in the local link will go to 'try_send' code branch, which is not expected.

Fixes: #77401

jukkar commented on include/zephyr/net/net_ip.h:
```diff
@@ -745,6 +745,39 @@ static inline bool net_ipv6_is_prefix(const uint8_t *addr1,
+ * @param outaddr Prefix masked IPv6 address (network address).
+ * @param length Prefix length (max length is 128).
+ */
+static inline void net_ipv6_addr_prefix_mask(const uint8_t *inaddr,
+		uint8_t *outaddr, uint8_t prefix_len)
```
Could you align the parameters like this
```
static inline void net_ipv6_addr_prefix_mask(const uint8_t *inaddr,
					     uint8_t *outaddr,
					     uint8_t prefix_len)
```

shrek-wang committed:
net: Add ipv6 prefix mask function

Use the IPv6 unicast address and the prefix-len to get the IPv6
network address.

Signed-off-by: Shrek Wang <shrek.wang@nxp.com>

jukkar commented on include/zephyr/net/net_ip.h:
```diff
@@ -745,6 +745,39 @@ static inline bool net_ipv6_is_prefix(const uint8_t *addr1,
 }
 
+
+/**
+ * @brief Get the IPv6 network address via the unicast address and the prefix mask.
+ *
+ * @param inaddr Unicast IPv6 address.
+ * @param outaddr Prefix masked IPv6 address (network address).
+ * @param length Prefix length (max length is 128).
+ */
```
Please place this function in a separate commit (we do not want to mix sample changes to other network changes in the same commit).

jukkar commented on samples/net/vlan/src/main.c:
```diff
@@ -63,6 +63,19 @@ static int setup_iface(struct net_if *iface, struct net_if *vlan,
 			return -EINVAL;
 		}
+
+		net_ipv6_addr_prefix_mask((uint8_t *)&addr6, (uint8_t *)&netaddr6,
+					CONFIG_NET_SAMPLE_IFACE_MY_IPV6_PREFIXLEN);
+
+		if (!net_if_ipv6_prefix_add(vlan, &netaddr6,
+					CONFIG_NET_SAMPLE_IFACE_MY_IPV6_PREFIXLEN,
+					(uint32_t)0xffffffff)) {
+			LOG_ERR("Cannot add %s with prefix_len %d to interface %p",
+					ipv6_addr,
+					CONFIG_NET_SAMPLE_IFACE_MY_IPV6_PREFIXLEN,
+					vlan);
```
Could you align the parameters in these functions

jukkar commented:
No need to keep this as a draft, the changes look fine and can be reviewed already.

shrek-wang committed:
samples: net: vlan: Add IPv6 prefix config to each vlan-iface

Without the IPv6-prefix set, the net_if_ipv6_addr_onlink() check
will always fail. The packets with DIP in the local link will go
to 'try_send' code branch, which is not expected.

Signed-off-by: Shrek Wang <shrek.wang@nxp.com>

jukkar commented on include/zephyr/net/net_ip.h:
```diff
@@ -745,6 +745,38 @@ static inline bool net_ipv6_is_prefix(const uint8_t *addr1,
 	return (addr1[bytes] & mask) == (addr2[bytes] & mask);
```
Please change the commit subject of the 1st one to `net: Add ipv6 prefix mask function`

jukkar commented on samples/net/vlan/Kconfig:
```diff
@@ -53,4 +53,12 @@ config NET_SAMPLE_IFACE3_VLAN_TAG
 	  Set VLAN (virtual LAN) tag (id) that is used in the sample
```
and the commit subject of the second one to `samples: net: vlan: Add IPv6 prefix config to each vlan-iface`

jukkar reviewed (commented):
The signed-off-by has error, the compliance checker says
```
1: UC2 Signed-off-by: must have a full name
```
for both commits.

shrek-wang commented:
All done. Thanks, @jukkar :)

jukkar commented:
Some fixed need for documentation (seen in CI)
```
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
WARNING: /home/runner/_work/zephyr/zephyr/include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
```

shrek-wang commented:
> Some fixes need for documentation (seen in CI)
> 
> ```
> WARNING: include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
> WARNING: include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
> WARNING: include/zephyr/net/net_ip.h:752: argument 'length' of command @param is not found in the argument list of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len)
> WARNING: include/zephyr/net/net_ip.h:752: The following parameter of net_ipv6_addr_prefix_mask(const uint8_t *inaddr, uint8_t *outaddr, uint8_t prefix_len) is not documented:
> ```

shrek-wang commented:
@jukkar , I've fixed the argument to the right one.
